### PR TITLE
[ENH] Added BaseDeepForecaster in forecasting/deep_learning

### LINKED ISSUE #2904: [ENH]  Add `BaseDeepForecaster` to `aeon.forecasting.deep_learning`
### Describe the feature or idea you want to propose

Introduce a new base class, `BaseDeepForecaster`, to the `aeon.forecasting.deep_learning` module. This class provides a clean foundation for implementing deep learning-based forecasting models using TensorFlow/Keras.

### Describe your proposed solution

Currently, `aeon` lacks a unified base class tailored for deep learning forecasting pipelines.

The `BaseDeepForecaster` is built on top of `BaseForecaster` while supporting key DL workflows. It can act as a foundation for models like Informer, TCN and DeepAR forcasters.

The proposed `BaseDeepForecaster` would:
- Inherit from `BaseForecaster`, passing `horizon` and `axis` via constructor.
- Define the abstract method `_build_model(input_shape)` for subclasses to implement specific architectures.
- Handle input conversion, sequence generation, model compilation, and training in `_fit`.
- Use the last observed window for forecasting in `_predict`.
- Support configuration via parameters: `horizon`, `window`, `batch_size`, `epochs`, `verbose`, `optimizer`, `loss`, and `random_state`.

### Describe alternatives you've considered, if relevant

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by lucifer4073**
#### Reference Issues/PRs

Fixes #2904.

#### What does this implement/fix? Explain your changes.

To lay the groundwork for future deep learning forecasters in aeon, BaseDeepForecaster class has been developed. This class abstracts common deep learning workflows for forecasting, such as input conversion, sequence generation, model compilation, training, and prediction.

#### Does your contribution introduce a new dependency? If yes, which one?

No.

#### Any other comments?

This modular base class will serve as the foundation for integrating advanced deep learning models like Informer, TCN, and DeepAR into the aeon toolkit.

##### For all contributions
- [X] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you **after** the PR has been merged.
- [X] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [ ] I've added the estimator/function to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).
- [ ] (OPTIONAL) I've added myself as a `__maintainer__` at the top of relevant files and want to be contacted regarding its maintenance. Unmaintained files may be removed. This is for the full file, and you should not add yourself if you are just making minor changes or do not want to help maintain its contents.

##### For developers with write access
- [ ] (OPTIONAL) I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.

### CODE REVIEW

File: aeon/forecasting/deep_learning/base.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+        for i in range(len(data) - self.window - self.horizon + 1):
+            X.append(data[i : (i + self.window)])
+            y.append(data[i + self.window : (i + self.window + self.horizon)])
+
+        X = np.array(X).reshape(-1, self.window, 1)
+        y = np.array(y).reshape(-1, self.horizon)
+        return X, y
+
+    @abstractmethod
+    def _build_model(self, input_shape):
```
Comment (hadifawaz1999) [inline aeon/forecasting/deep_learning/base.py]: you dont have to make this function private just call it `build_model`

File: aeon/forecasting/deep_learning/base.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+from abc import abstractmethod
+
+import numpy as np
+import pandas as pd
+
+from aeon.forecasting.base import BaseForecaster
+
+
+class BaseDeepForecaster(BaseForecaster):
```
Comment (hadifawaz1999) [inline aeon/forecasting/deep_learning/base.py:18]: the base class should have the following tags

```python
_tags = {
        "capability:horizon": False,
        "capability:exogenous": False,
        "algorithm_type": "deeplearning",
        "non_deterministic": True,
        "cant_pickle": True,
        "python_dependencies": "tensorflow",
    }
```

Comment (hadifawaz1999) [reply aeon/forecasting/deep_learning/base.py:18]: the rest of the tags will be inherited from base forecaster, but these are deep learning specific for now

File: aeon/forecasting/deep_learning/base.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+        Returns
+        -------
+        self : BaseDeepForecaster
+            Returns an instance of self.
+        """
+        import tensorflow as tf
+
+        # Set random seed for reproducibility
+        if self.random_state is not None:
```
Comment (hadifawaz1999) [inline aeon/forecasting/deep_learning/base.py]: random state should be done like that:
```python
rng = check_random_state(self.random_state)
self.random_state_ = rng.randint(0, np.iinfo(np.int32).max)
tf.keras.utils.set_random_seed(self.random_state_)
```

File: aeon/forecasting/deep_learning/base.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+        Batch size for training the model.
+    epochs : int, default=100
+        Number of epochs to train the model.
+    verbose : int, default=0
+        Verbosity mode (0, 1, or 2).
+    optimizer : str or tf.keras.optimizers.Optimizer, default='adam'
+        Optimizer to use for training.
+    loss : str or tf.keras.losses.Loss, default='mse'
+        Loss function for training.
+    random_state : int, default=None
```
Comment (hadifawaz1999) [inline aeon/forecasting/deep_learning/base.py]: random_state description should be like this:

random_state : int, RandomState instance or None, default=None
        If `int`, random_state is the seed used by the random number generator;
        If `RandomState` instance, random_state is the random number generator;
        If `None`, the random number generator is the `RandomState` instance used
        by `np.random`.
        Seeded random number generation can only be guaranteed on CPU processing,
        GPU processing will be non-deterministic.

File: aeon/forecasting/deep_learning/base.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+        Axis along which to apply the forecaster.
+        Default is 0 for univariate time series.
+    """
+
+    def __init__(
+        self,
+        horizon=1,
+        window=10,
+        batch_size=32,
+        epochs=100,
```
Comment (hadifawaz1999) [inline aeon/forecasting/deep_learning/base.py]: n_epochs not epochs, same in docstring

File: aeon/forecasting/deep_learning/base.py [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+            X_train,
+            y_train,
+            batch_size=self.batch_size,
+            epochs=self.n_epochs,
+            verbose=self.verbose,
+        )
+        self.last_window_ = y_inner[-self.window :]
+        return self
+
+    def _predict(self, y=None, X=None):
```
Comment (TonyBagnall) [inline aeon/forecasting/deep_learning/base.py]: I prefer exog to X. This appeared in the wrong place 

def _predict(self, y, exog=None)

**Review by TonyBagnall — CHANGES_REQUESTED** [review]

File: aeon/forecasting/deep_learning/base.py (hunk 6 above)
Comment (hadifawaz1999) [reply aeon/forecasting/deep_learning/base.py]: didnt notice that good catch

**Review by MatthewMiddlehurst — COMMENTED** [review]
No comments other than those Tony has posted for now. Try to match the syntax and terminology used in the module. Consider of things such as the new mixins as well, they may not be required here specifically but is there anything special for deep learners that we need to think about.

**Comment by hadifawaz1999** [comment]
> No comments other than those Tony has posted for now. Try to match the syntax and terminology used in the module. Consider of things such as the new mixins as well, they may not be required here specifically but is there anything special for deep learners that we need to think about.

i would keep mixins for models themselves, as not all deep learners follow the same setup

File: aeon/forecasting/deep_learning/base.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+
+    def __init__(
+        self,
+        horizon=1,
+        window=10,
+        verbose=0,
+        callbacks=None,
+        axis=0,
+        last_file_name="last_model",
+        save_best_model=False,
```
Comment (hadifawaz1999) [inline aeon/forecasting/deep_learning/base.py]: the `save_best_model` parameter doesnt have to be in base, its not needed

File: aeon/forecasting/deep_learning/base.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+        callbacks_list : list
+            List of callbacks to be used during training.
+        """
+        callbacks_list = []
+        if self.callbacks is not None:
+            if isinstance(self.callbacks, list):
+                callbacks_list.extend(self.callbacks)
+            else:
+                callbacks_list.append(self.callbacks)
+        if self.save_best_model:
```
Comment (hadifawaz1999) [inline aeon/forecasting/deep_learning/base.py]: the model checkpoint should be added regardless if the user chooses save_best_model. The save_best_model parameter will be used in fit, if its true then fit will not delete the file saved by the model checkpoint, and if its false it will delete it. This will be implemented in fit for each model, so you can remove the if condition because you should always add the model checkpoint

File: aeon/forecasting/deep_learning/base.py [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+        file_path : str, default="./"
+            The directory where the model will be saved.
+
+        Returns
+        -------
+        None
+        """
+        if self.model_ is None:
+            raise ValueError("No model to save. Please fit the model first.")
+        self.model_.save(file_path + self.last_file_name + ".keras")
```
Comment (hadifawaz1999) [inline aeon/forecasting/deep_learning/base.py]: can you replace ```file_path + self.last_file_name + ".keras"``` by ```os.path.join(file_path, self.last_file_name + ".keras")```

File: aeon/forecasting/deep_learning/base.py [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+        The window size for creating input sequences.
+    verbose : int, default=0
+        Verbosity mode (0, 1, or 2).
+    callbacks : list of tf.keras.callbacks.Callback or None, default=None
+        List of Keras callbacks to be applied during training.
+    axis : int, default=0
+        Axis along which to apply the forecaster.
+    last_file_name : str, default="last_model"
+        The name of the file of the last model, used for saving models.
+    save_best_model : bool, default=False
```
Comment (hadifawaz1999) [inline aeon/forecasting/deep_learning/base.py]: remove save_best_model from docs too

File: aeon/forecasting/deep_learning/base.py [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+
+    This class provides a foundation for deep learning-based forecasting models,
+    handling data preprocessing, model training, and prediction with enhanced
+    capabilities for callbacks, model saving/loading, and efficiency.
+
+    Parameters
+    ----------
+    horizon : int, default=1
+        Forecasting horizon, the number of steps ahead to predict.
+    window : int, default=10
```
Comment (TonyBagnall) [inline aeon/forecasting/deep_learning/base.py]: is this default not quite small?

File: aeon/forecasting/deep_learning/base.py [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+        The window size for creating input sequences.
+    verbose : int, default=0
+        Verbosity mode (0, 1, or 2).
```
Comment (TonyBagnall) [inline aeon/forecasting/deep_learning/base.py:32]: what is the difference between these modes?

File: aeon/forecasting/deep_learning/base.py [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+    """
+
+    _tags = {
+        "capability:horizon": True,
+        "capability:exogenous": False,
+        "algorithm_type": "deeplearning",
+        "non_deterministic": True,
+        "cant_pickle": True,
+        "python_dependencies": "tensorflow",
+        "capability:multivariate": True,
```
Comment (TonyBagnall) [inline aeon/forecasting/deep_learning/base.py]: I dont think we have any general test for multivariate yet, it might break existing tests since it will return an array not a scalar?

File: aeon/forecasting/deep_learning/base.py [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+    file_path : str, default="./"
+        Directory path where models will be saved.
+
+    Attributes
+    ----------
+    model_ : tf.keras.Model or None
+        The fitted Keras model.
+    history_ : tf.keras.callbacks.History or None
+        Training history containing loss and metrics.
+    last_window_ : np.ndarray or None
```
Comment (TonyBagnall) [inline aeon/forecasting/deep_learning/base.py:48]: can these be None after fit? I personally prefer declaration in fit rather than the constructor, but not sure

File: aeon/forecasting/deep_learning/base.py [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+        """Fit the model."""
+        pass
+
+    def _predict(self, y, exog=None):
+        """Predict using the model."""
+        pass
+
+    def _forecast(self, y, exog=None):
+        """Forecast values for time series X."""
+        y = self._preprocess_series(y, 1, True)
```
Comment (TonyBagnall) [inline aeon/forecasting/deep_learning/base.py]: passing axis 1 here, should it not be self.axis?

File: aeon/forecasting/deep_learning/base.py [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+        ----------
+        input_shape : tuple
+            Shape of input data.
+
+        Returns
+        -------
+        model : tf.keras.Model
+            Compiled Keras model.
+        """
+        pass
```
Comment (TonyBagnall) [inline aeon/forecasting/deep_learning/base.py:199]: dont  think we want pass here, use ellipsis ...

**Review by TonyBagnall — CHANGES_REQUESTED** [review]

File: aeon/forecasting/deep_learning/base.py (hunk 12 above)
Comment (lucifer4073) [reply aeon/forecasting/deep_learning/base.py:32]: 0 refers to no output.
1 means a progress bar.
2 means one line per epoch.
These are the standard values that verbose takes in `tf.keras`.

**Review by hadifawaz1999 — APPROVED** [review]
LGTM! @TonyBagnall

File: aeon/forecasting/deep_learning/base.py [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+        self.last_file_name = last_file_name
+        self.file_path = file_path
+
+        self.model_ = None
+        self.history_ = None
+        self.last_window_ = None
+
+        super().__init__(horizon=horizon, axis=axis)
+
+    def _fit(self, y, exog=None):
```
Comment (TonyBagnall) [inline aeon/forecasting/deep_learning/base.py]: the default fit in base just returns self. Is the intention to require it for BaseDeepForecasters?

File: aeon/forecasting/deep_learning/base.py [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+        """Fit the model."""
+        ...
+
+    def _predict(self, y, exog=None):
```
Comment (TonyBagnall) [inline aeon/forecasting/deep_learning/base.py]: this is the same as the base class predict, so isnt needed I dont think

File: aeon/forecasting/deep_learning/base.py [hunk 19]
```diff
@@ -0,0 +1,229 @@
...
+        self.model_ = None
+        self.history_ = None
+        self.last_window_ = None
+
+        super().__init__(horizon=horizon, axis=axis)
+
+    def _fit(self, y, exog=None):
+        """Fit the model."""
+        ...
+
```
Comment (TonyBagnall) [inline aeon/forecasting/deep_learning/base.py]: abstract methods should be annotated abstract

File: aeon/forecasting/deep_learning/tests/test_base.py [hunk 20]
```diff
@@ -0,0 +1,46 @@
+"""Test file for BaseDeepForecaster."""
+
+import pytest
+
+from aeon.forecasting.deep_learning.base import BaseDeepForecaster
+from aeon.utils.validation._dependencies import _check_soft_dependencies
+
+
+class DummyDeepForecaster(BaseDeepForecaster):
```
Comment (TonyBagnall) [inline aeon/forecasting/deep_learning/tests/test_base.py:9]: doesnt implement abstract methods

**Review by TonyBagnall — CHANGES_REQUESTED** [review]
